Sources/Core/FBApplicationDelegate.swift

```diff
@@ -0,0 +1,329 @@
+        ) {
+        finished = true
+      }
+    }
+
+    return finished
+  }
+
+  // gets called from a notification that then passed the object which is the appliction to the observers
+  func applicationDidEnterBackground(_ notification: Notification) {
+    state = .background
+
+    observers.forEach { observer in
+      guard let application = notification.object as? UIApplication else {
+        return
+      }
+
+      observer.applicationObserving.applicationDidEnterBackground(application)
+    }
+  }
+
+  func applicationDidBecomeActive(_ notification: Notification) {
+    state = .active
+
+    if settings.isAutoLogAppEventsEnabled {
+      appEventsLogger.activateApp()
+    }
+
+    observers.forEach { observer in
+      guard let application = notification.object as? UIApplication else {
+        return
+      }
+
+      observer.applicationObserving.applicationDidBecomeActive(application)
+    }
+  }
+
+  private func logAppEventIfPresent(in url: URL) {
+    guard let components = URLComponents(url: url, resolvingAgainstBaseURL: false),
+      let queryItems = components.queryItems,
+      let eventData = queryItems.decodeFromItem(
+        withName: "al_applink_data",
+        AppLinkEventData.self
+      )
+      else {
+        return
+    }
+
+    var parameters = [String: String]()
+
+    if let urlString = eventData.targetURLString,
+      let targetURL = URL(string: urlString) {
+      parameters.updateValue(targetURL.absoluteString, forKey: AppEventsLogger.LinkEventKeys.targetURL)
+
+      if let host = targetURL.host {
+        parameters.updateValue(host, forKey: AppEventsLogger.LinkEventKeys.targetURLHost)
+      }
+    }
+
+    if let referralTargetUrlString = eventData.refererData?.targetURLString {
+      parameters.updateValue(referralTargetUrlString, forKey: AppEventsLogger.LinkEventKeys.referralTargetURL)
+    }
+
+    if let referralURLString = eventData.refererData?.url {
+      parameters.updateValue(referralURLString, forKey: AppEventsLogger.LinkEventKeys.referralURL)
+    }
+
+    if let appName = eventData.refererData?.appName {
+      parameters.updateValue(appName, forKey: AppEventsLogger.LinkEventKeys.appName)
+    }
+
+    parameters.updateValue(url.absoluteString, forKey: AppEventsLogger.LinkEventKeys.inputURL)
+
+    if let scheme = url.scheme {
+      parameters.updateValue(scheme, forKey: AppEventsLogger.LinkEventKeys.inputURLScheme)
+    }
+
+    appEventsLogger.logInternalEvent(
+      eventName: "fb_al_inbound",
+      parameters: parameters,
+      isImplicitlyLogged: true
+    )
+  }
+
+  private func logSDKInitialization() {
+    // TODO: Implement collecting the actual data
+
+    appEventsLogger.logInternalEvent(
+      eventName: "fb_sdk_initialize",
+      parameters: [:],
+      isImplicitlyLogged: false
+    )
+  }
+
+  enum Errors: FBError {
+    case invalidArgument
+
+    var developerMessage: String {
+      return """
+Expected 'sourceApplication' to be non-nil and non-empty.
```

Comment:
Indentation is measured against the bottom 3 quotes. Proper indentation is acceptable and encouraged.

```suggestion
      Expected 'sourceApplication' to be non-nil and non-empty.
      Please verify you are passing in 'sourceApplication' from your app delegate and not the UIApplication parameter.
      If your app delegate implements iOS 9's application:openURL:options:, you should pass in
      options[OpenURLOptionsKey.sourceApplication].
      """
```

Reply:
Fixed in later commit.